internal/engine/table_iterator.go

```diff
@@ -3,43 +3,71 @@ package engine
 import (
 	"fmt"
 
+	"github.com/xqueries/xdb/internal/engine/page"
+
+	"github.com/xqueries/xdb/internal/engine/dbfs"
 	"github.com/xqueries/xdb/internal/engine/profile"
-	"github.com/xqueries/xdb/internal/engine/storage/page"
+	"github.com/xqueries/xdb/internal/engine/schema"
 	"github.com/xqueries/xdb/internal/engine/table"
 )
 
 type tableRowIterator struct {
 	profiler *profile.Profiler
 
-	cols     []table.Col
-	dataPage *page.Page
+	schema *schema.Schema
+	data   *dbfs.PagedFile
+
+	pages            []page.ID
+	currentPageIndex int
+	currentPage      *page.Page
 
-	index int
-	slots []page.Slot
+	slots       []page.Slot
+	currentSlot int
 }
 
-func newTableRowIterator(profiler *profile.Profiler, cols []table.Col, dataPage *page.Page) *tableRowIterator {
+func newTableRowIterator(profiler *profile.Profiler, schema *schema.Schema, data *dbfs.PagedFile) *tableRowIterator {
 	return &tableRowIterator{
 		profiler: profiler,
-		cols:     cols,
-		dataPage: dataPage,
+		schema:   schema,
+		data:     data,
+		pages:    data.Pages(),
 	}
 }
 
 // Next returns the next row of this table iterator.
 func (i *tableRowIterator) Next() (table.Row, error) {
 	i.profiler.Enter("next row").Exit()
 
-	if i.slots == nil {
-		i.slots = i.dataPage.OccupiedSlots()
+	if len(i.pages) == 0 {
+		return table.Row{}, table.ErrEOT
 	}
-	if i.index >= len(i.slots) {
+
+start:
```

Comment:
Can you briefly explain the flow happening inside this label?

Reply:
Because there are many conditionals, those are best documented instead of thinking hard next time to figure out what and why it was done for :))